File: tasks/native-runtime/mindmap_parity_probe.py
"""create_mindmap parity probe, Python side.

Pins label cleaning, node normalization, layout, and the SVG bytes of
`create_mindmap` against `deepseek_policy::mindmaps`. File ids are injected so
the store envelope is comparable; the SVG itself does not contain the id.

Usage::

    python tasks/native-runtime/mindmap_parity_probe.py > python.json
    cd rust && cargo run -p deepseek-policy --example mindmap_parity_probe > ../rust.json
    diff <(tr -d '\\r' < python.json) <(tr -d '\\r' < rust.json)
"""

from __future__ import annotations

import json
import os
import shutil
import sys
import tempfile
from pathlib import Path
from typing import Any
from unittest.mock import patch

REPO = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(REPO))

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8")

from deepseek_infra.core.errors import AppError  # noqa: E402
from deepseek_infra.infra.tool_runtime import generated_files, mindmaps  # noqa: E402

SAMPLE = [
    {
        "label": "Market analysis",
        "children": [
            {"label": "User profile", "children": []},
            {"label": "Competition", "children": [{"label": "Pricing", "children": []}]},
        ],
    },
    {
        "label": "Product strategy",
        "children": [
            {"label": "Core features", "children": []},
            {"label": "Launch rhythm", "children": []},
        ],
    },
]


def error_view(exc: AppError) -> dict[str, Any]:
    return {"error": str(exc), "code": exc.code.value, "status": exc.status}


def make(title: str, nodes: Any, subtitle: str, root: Path, file_id: str = "a" * 32) -> dict[str, Any]:
    generated_files.GENERATED_DIR = root / ".generated"
    with patch("secrets.token_hex", return_value=file_id):
        result = mindmaps.create_mindmap(title, nodes, subtitle=subtitle)
    path = generated_files.resolve_generated_file(result["fileId"])
    assert path is not None
    result = dict(result)
    result["svg"] = path.read_text(encoding="utf-8")
    return result


def main() -> int:
    out: dict[str, Any] = {}
    root = Path(tempfile.mkdtemp(prefix="mindmap-parity-"))
    os.environ["DEEPSEEK_INFRA_ROOT"] = str(root)

    try:
        make("", SAMPLE, "", root)
        out["empty-title"] = "ok"
    except AppError as exc:
        out["empty-title"] = error_view(exc)
    try:
        make("Empty", [], "", root)
        out["empty-nodes"] = "ok"
    except AppError as exc:
        out["empty-nodes"] = error_view(exc)

    grown = make("Growth plan", SAMPLE, "2026", root, file_id="b" * 32)
    out["growth::format"] = grown["format"]
    out["growth::nodeCount"] = grown["nodeCount"]
    out["growth::title"] = grown["title"]
    out["growth::outline"] = grown["outline"]
    out["growth::filename"] = grown["filename"]
    out["growth::downloadUrl"] = grown["downloadUrl"]
    out["growth::svg"] = grown["svg"]

    escaped = make("<Title>", [{"label": "A & B", "children": [{"title": "Child"}]}], "sub", root, file_id="c" * 32)
    out["escape::svg"] = escaped["svg"]

    aliases = make(
        "Aliases",
        [None, "", "one", {"title": "two"}, {"name": "three"}],
        "",
        root,
        file_id="d" * 32,
    )
    out["aliases::outline"] = aliases["outline"]
    out["aliases::nodeCount"] = aliases["nodeCount"]

    shutil.rmtree(root, ignore_errors=True)
    json.dump(out, sys.stdout, ensure_ascii=False, indent=2, sort_keys=True)
    sys.stdout.write("\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/crates/deepseek-policy/examples/mindmap_parity_probe.rs
//! create_mindmap parity probe, Rust side.
//!
//! Usage::
//!
//!     python tasks/native-runtime/mindmap_parity_probe.py > python.json
//!     cd rust && cargo run -p deepseek-policy --example mindmap_parity_probe > ../rust.json
//!     diff <(tr -d '\r' < python.json) <(tr -d '\r' < rust.json)

use std::fs;
use std::path::{Path, PathBuf};

use deepseek_policy::app_error::AppError;
use deepseek_policy::entropy::Entropy;
use deepseek_policy::generated_files;
use deepseek_policy::mindmaps::create_mindmap;
use serde_json::{Map, Value, json};

struct FixedEntropy {
    id: String,
}

impl Entropy for FixedEntropy {
    fn new_id(&self) -> Result<String, AppError> {
        Ok(self.id.chars().take(16).collect())
    }
    fn new_file_id(&self) -> Result<String, AppError> {
        Ok(self.id.clone())
    }
    fn now_millis(&self) -> i64 {
        0
    }
}

fn error_view(error: &AppError) -> Value {
    json!({"error": error.message, "code": error.code, "status": error.status})
}

fn make(title: &str, nodes: &Value, subtitle: &str, root: &Path, file_id: &str) -> Value {
    let entropy = FixedEntropy {
        id: file_id.to_string(),
    };
    match create_mindmap(title, nodes, subtitle, root, &entropy, 1_700_000_000.0) {
        Ok(mut result) => {
            let id = result["fileId"].as_str().unwrap().to_string();
            let path = generated_files::resolve_generated_file(root, &id).unwrap();
            let svg = fs::read_to_string(path).unwrap();
            result
                .as_object_mut()
                .unwrap()
                .insert("svg".to_string(), json!(svg));
            result
        }
        Err(error) => error_view(&error),
    }
}

fn main() {
    let mut out = Map::new();
    let scratch: PathBuf =
        std::env::temp_dir().join(format!("mindmap-parity-{}", std::process::id()));
    let _ = fs::remove_dir_all(&scratch);
    fs::create_dir_all(&scratch).unwrap();

    let sample = json!([
        {
            "label": "Market analysis",
            "children": [
                {"label": "User profile", "children": []},
                {"label": "Competition", "children": [{"label": "Pricing", "children": []}]}
            ]
        },
        {
            "label": "Product strategy",
            "children": [
                {"label": "Core features", "children": []},
                {"label": "Launch rhythm", "children": []}
            ]
        }
    ]);

    out.insert(
        "empty-title".to_string(),
        make("", &sample, "", &scratch, &"a".repeat(32)),
    );
    out.insert(
        "empty-nodes".to_string(),
        make("Empty", &json!([]), "", &scratch, &"a".repeat(32)),
    );

    let grown = make("Growth plan", &sample, "2026", &scratch, &"b".repeat(32));
    out.insert("growth::format".to_string(), grown["format"].clone());
    out.insert("growth::nodeCount".to_string(), grown["nodeCount"].clone());
    out.insert("growth::title".to_string(), grown["title"].clone());
    out.insert("growth::outline".to_string(), grown["outline"].clone());
    out.insert("growth::filename".to_string(), grown["filename"].clone());
    out.insert(
        "growth::downloadUrl".to_string(),
        grown["downloadUrl"].clone(),
    );
    out.insert("growth::svg".to_string(), grown["svg"].clone());

    let escaped = make(
        "<Title>",
        &json!([{"label": "A & B", "children": [{"title": "Child"}]}]),
        "sub",
        &scratch,
        &"c".repeat(32),
    );
    out.insert("escape::svg".to_string(), escaped["svg"].clone());

    let aliases = make(
        "Aliases",
        &json!([null, "", "one", {"title": "two"}, {"name": "three"}]),
        "",
        &scratch,
        &"d".repeat(32),
    );
    out.insert("aliases::outline".to_string(), aliases["outline"].clone());
    out.insert(
        "aliases::nodeCount".to_string(),
        aliases["nodeCount"].clone(),
    );

    let _ = fs::remove_dir_all(&scratch);
    let mut encoded = serde_json::to_string_pretty(&Value::Object(out)).expect("serialize");
    encoded.push('\n');
    print!("{encoded}");
}
